jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/ParallelGraphWriter.java

```diff
@@ -0,0 +1,321 @@
+            buffer.order(java.nio.ByteOrder.BIG_ENDIAN);
+            return buffer;
+        });
+    }
+
+    /**
+     * Writes all L0 node records in parallel using asynchronous file I/O.
+     * Records are written in order to maintain index correctness.
+     *
+     * @param ordinalMapper maps between old and new ordinals
+     * @param inlineFeatures the inline features to write
+     * @param featureStateSuppliers suppliers for feature state
+     * @param baseOffset the file offset where L0 records start
+     * @throws IOException if an IO error occurs
+     */
+    public void writeL0Records(OrdinalMapper ordinalMapper,
+                               List<Feature> inlineFeatures,
+                               Map<FeatureId, IntFunction<Feature.State>> featureStateSuppliers,
+                               long baseOffset) throws IOException {
+        // Build all records in parallel
+        List<Future<NodeRecordTask.Result>> futures = buildRecords(
+                ordinalMapper, inlineFeatures, featureStateSuppliers, baseOffset);
+
+        // Write results in order using async I/O
+        writeRecordsAsync(futures);
+    }
+
+    /**
+     * Builds all node records in parallel using the worker thread pool.
+     * Each worker thread obtains its own graph view and buffer from thread-local storage.
+     *
+     * @param ordinalMapper maps between old and new ordinals
+     * @param inlineFeatures the inline features to write
+     * @param featureStateSuppliers suppliers for feature state
+     * @param baseOffset the file offset where L0 records start
+     * @return list of futures for all record building tasks
+     */
+    private List<Future<NodeRecordTask.Result>> buildRecords(
+            OrdinalMapper ordinalMapper,
+            List<Feature> inlineFeatures,
+            Map<FeatureId, IntFunction<Feature.State>> featureStateSuppliers,
+            long baseOffset) {
+        int maxOrdinal = ordinalMapper.maxOrdinal();
+        List<Future<NodeRecordTask.Result>> futures = new ArrayList<>(maxOrdinal + 1);
+
+        for (int newOrdinal = 0; newOrdinal <= maxOrdinal; newOrdinal++) {
+            final int ordinal = newOrdinal;
+            final long fileOffset = baseOffset + (long) ordinal * recordSize;
+
+            Future<NodeRecordTask.Result> future = executor.submit(() -> {
+                var view = viewPerThread.get();
+                var buffer = bufferPerThread.get();
+
+                var task = new NodeRecordTask(
+                        ordinal,
+                        ordinalMapper,
+                        graph,
+                        view,
+                        inlineFeatures,
+                        featureStateSuppliers,
+                        recordSize,
+                        fileOffset,
+                        buffer
+                );
+
+                return task.call();
+            });
+
+            futures.add(future);
+        }
```

Comment:
IIUC, this doesn't have any back pressure mechanism, and we need that to prevent excessive memory utilization. I think we might want to consider an implementation that uses semaphores to manage concurrency.